File: frontend/src/features/system-health/schemas.ts
import { z } from "zod";

export const SystemHealthMetricsSchema = z.object({
  totalAccounts: z.number().int().nonnegative().nullable().optional(),
  activeAccounts: z.number().int().nonnegative().nullable().optional(),
  unavailableAccounts: z.number().int().nonnegative().nullable().optional(),
  unavailableRatio: z.number().nonnegative().nullable().optional(),
  requestCount: z.number().int().nonnegative().nullable().optional(),
  rateLimitRatio: z.number().nonnegative().nullable().optional(),
  projectedExhaustionAt: z.string().datetime({ offset: true }).nullable().optional(),
  riskLevel: z.enum(["warning", "danger", "critical"]).nullable().optional(),
});

export const SystemHealthAlertSchema = z.object({
  code: z.string(),
  severity: z.enum(["warning", "critical"]),
  title: z.string(),
  message: z.string(),
  href: z.string(),
  metrics: SystemHealthMetricsSchema.nullable().optional(),
});

export const SystemHealthResponseSchema = z.object({
  status: z.enum(["healthy", "warning", "critical"]),
  updatedAt: z.string().datetime({ offset: true }),
  alert: SystemHealthAlertSchema.nullable(),
});

export type SystemHealthMetrics = z.infer<typeof SystemHealthMetricsSchema>;
export type SystemHealthAlert = z.infer<typeof SystemHealthAlertSchema>;
export type SystemHealthResponse = z.infer<typeof SystemHealthResponseSchema>;

File: app/modules/system_health/schemas.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import Field

from app.modules.shared.schemas import DashboardModel

SystemHealthStatus = Literal["healthy", "warning", "critical"]
SystemHealthSeverity = Literal["warning", "critical"]


class SystemHealthMetrics(DashboardModel):
    total_accounts: int | None = Field(default=None)
    active_accounts: int | None = Field(default=None)
    unavailable_accounts: int | None = Field(default=None)
    unavailable_ratio: float | None = Field(default=None)
    request_count: int | None = Field(default=None)
    rate_limit_ratio: float | None = Field(default=None)


class SystemHealthAlert(DashboardModel):
    code: str
    severity: SystemHealthSeverity
    title: str
    message: str
    href: str
    metrics: SystemHealthMetrics | None = None


class SystemHealthResponse(DashboardModel):
    status: SystemHealthStatus
    updated_at: datetime
    alert: SystemHealthAlert | None = None
